pkg/operator/ceph/object/bucket/controller.go

```diff
@@ -97,6 +97,19 @@ func add(ctx context.Context, mgr manager.Manager, r reconcile.Reconciler) error
 		return err
 	}
 
+	// Watch for CephObjectStore spec change
+	objectStoreKind := source.Kind(mgr.GetCache(),
+		&cephv1.CephObjectStore{TypeMeta: metav1.TypeMeta{Kind: "CephObjectStore", APIVersion: v1.SchemeGroupVersion.String()}})
+	err = c.Watch(objectStoreKind, handler.EnqueueRequestsFromMapFunc(handler.MapFunc(func(context context.Context, obj client.Object) []reconcile.Request {
+		objectStore, _ := obj.(*cephv1.CephObjectStore)
+		req := reconcile.Request{NamespacedName: types.NamespacedName{Namespace: objectStore.Namespace}}
+		return []reconcile.Request{req}
+	}),
```

Comment:
This map function doesn't do the right thing. The map function is supposed to transform requests for one resource into a list of resources of another type. In this case, we would theoretically transform a reconcile to a CephObjectStore to the full list of OBCs affected by the CephObjectStore change.

This map function seems to translate a CephObjectStore change into a different CephObjectStore that lacks name and only contains a namespace. I especially don't think we should be requeuing the request in a way that removes the name.

However, it's possible we might not want a map function. I think that may be determined based on where the map function is called in relation to the predicate. Is the map function called after the predicate processes? If so, then we can use a map function to "fan out" the object store to all affected OBCs. 

But if the map function happens before the predicate, then the predicate would never receive objectstore resources, and we don't have a way of detecting objectstore status changes to lead us to reconcile specific OBCs. In that case, we have to reconcile cephobjectstores and then, in that reconcile loop, also reconcile each OBC. Yuval did a similar thing with his work in the bucket notification controller.

Reply:
I clarified this question with controller-runtime devs. The source is processed by the predicate, then the map function (enqueue step), and then that is passed immediately to the reconciler. This is the most ideal case for us here.

That means that the predicate in this PR is on the right track, but the map function needs updated to list all OBCs and return a list of all of them that are affected by the CephObjectStore change.

Please also define the handler function separately so that the code is easier to read, as is done below:

https://github.com/rook/rook/blob/5bff860380f34d1aa1ee9dceb9bcb2ed5d64da58/pkg/operator/ceph/cluster/controller.go#L167-L188

Reply:
Yes correct first is the predicate called and then the even handler, thanks for sharing this document earlier https://book-v1.book.kubebuilder.io/beyond_basics/controller_watches.html it is very detailed one.
And thanks for confirming with controller-runtime devs.

And I have tested the changes it works like that...